libsolidity/experimental/ast/TypeSystemHelper.cpp

```diff
@@ -0,0 +1,404 @@
+{
+	switch (_sort.classes.size())
+	{
+	case 0:
+		return "()";
+	case 1:
+		return typeSystem.typeClassName(*_sort.classes.begin());
+	default:
+	{
+		std::stringstream stream;
+		stream << "(";
+		for (auto typeClass: _sort.classes | ranges::views::drop_last(1))
+			stream << typeSystem.typeClassName(typeClass) << ", ";
+		stream << typeSystem.typeClassName(*_sort.classes.rbegin()) << ")";
+		return stream.str();
+	}
+	}
+}
+
+std::string TypeEnvironmentHelpers::canonicalTypeName(Type _type) const
+{
+	return visit(util::GenericVisitor{
+		[&](TypeConstant _type) -> std::string {
+			std::stringstream stream;
+			stream << env.typeSystem().constructorInfo(_type.constructor).canonicalName;
+			if (!_type.arguments.empty())
+			{
+				stream << "$";
+				for (auto type: _type.arguments | ranges::views::drop_last(1))
+					stream << canonicalTypeName(type) << "$";
+				stream << canonicalTypeName(_type.arguments.back());
+				stream << "$";
+			}
+			return stream.str();
+		},
+		[](TypeVariable) -> std::string {
+			solAssert(false);
+		},
+		[](std::monostate) -> std::string {
+			solAssert(false);
+		},
+	}, env.resolve(_type));
+}
+
+std::string TypeEnvironmentHelpers::typeToString(Type const& _type) const
+{
+	std::map<TypeConstructor, std::function<std::string(std::vector<Type>)>> formatters{
+		{env.typeSystem().constructor(PrimitiveType::Function), [&](auto const& _args) {
+			solAssert(_args.size() == 2);
+			return fmt::format("{} -> {}", typeToString(_args.front()), typeToString(_args.back()));
+		}},
+		{env.typeSystem().constructor(PrimitiveType::Unit), [&](auto const& _args) {
+			solAssert(_args.size() == 0);
+			return "()";
+		}},
+		{env.typeSystem().constructor(PrimitiveType::Pair), [&](auto const& _arguments) {
+			auto tupleTypes = TypeSystemHelpers{env.typeSystem()}.destTupleType(_arguments.back());
+			std::string result = "(";
+			result += typeToString(_arguments.front());
+			for (auto type: tupleTypes)
+				result += ", " + typeToString(type);
+			result += ")";
+			return result;
+		}},
+	};
+	return std::visit(util::GenericVisitor{
+		[&](TypeConstant const& _type) {
+			if (auto* formatter = util::valueOrNullptr(formatters, _type.constructor))
+				return (*formatter)(_type.arguments);
+			std::stringstream stream;
+			stream << env.typeSystem().constructorInfo(_type.constructor).name;
+			if (!_type.arguments.empty())
+			{
+				stream << "(";
+				for (auto type: _type.arguments | ranges::views::drop_last(1))
+					stream << typeToString(type) << ", ";
+				stream << typeToString(_type.arguments.back());
+				stream << ")";
+			}
+			return stream.str();
+		},
+		[&](TypeVariable const& _type) {
+			std::stringstream stream;
+			std::string varName;
+			size_t index = _type.index();
+			varName += static_cast<char>('a' + (index%26));
+			while (index /= 26)
+				varName += static_cast<char>('a' + (index%26));
+			reverse(varName.begin(), varName.end());
+			stream << '\'' << varName;
```

Comment:
Is there an established convention for naming fixed free type variables? Is prefixing them with `"` a good idea?

Reply:
Change generic ones to be prefixed by ``?`` instead of ``'`` and use ``'`` for fixed ones.